### ISSUE TITLE
`reportOptionalOperand` has no effect in RHS

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1053]
**Describe the bug**
With strict type checking or `reportOptionalOperand` enabled, optionals do not produce errors when they are on the RHS of the operator.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
`1 + bar()` would produce an error.

**Screenshots or Code**
```python
# pyright: strict

from typing import Optional

# class int:
#     def __add__(self, x: int) -> int: ...

def bar() -> Optional[int]: ...

bar() + 1  # Operator "+" not supported for "None"
1 + bar()  # No error; TypeError at runtime if bar returns "None"
```

**VS Code extension or command-line**
Extension, 1.1.73

**Comment by erictraut** [comment]
Thanks for the bug report.

Despite its name, the `reportOptionalOperand` applies only to the left-hand operand in a binary operation. The left-hand operand generally provides the magic method that defines what types are permitted for the right-hand operand. 

You are correct that there were missing type checks. In particular, if the right-hand operand was a union, an error was generated only if _none_ of the subtypes within the union were permitted. It should emit an error if _any_ of the subtypes are incompatible. I've updated the code, and the fix will be in the next release.

Incidentally, when I added this stricter checking, it uncovered about a dozen _real_ bugs in my team's Python code base!

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.74, which I just published. It will also be included in the next release of Pylance.
